### PR TITLE
Evaluator changes for helper call

### PR DESCRIPTION
**Pull request by r30shah**
This PR consists of following changes. 
1. Correcting argument order of VM's JIT helpers to match proper C linkage which was previously following private linkage for historical reasons. x86 has already fixed this and in this commit we are correcting argument order on IBM Z.
2. Corresponding changes in `S390CHelperLinkage` to load argument register in correct order. 
3. We do not need to construct a new linkage every time we want to generate a helper call so use `getLinkage(TR_CHelper)` instead of `createLinkage(TR_CHelper)` in evaluators to generate helper call.
4. For 64-bit VM when we pass arguments to JIT helper for new allocations, we need to sign extend type/size child node of newarray or anewarray node as C helper linkage uses whole 64-bit register and evaluating 32-bit child of node only changes lower half of the register. This commit creates a call node `acall` for generating helper call sequence for new allocation node with 64-bit extended child if needed.

### CODE REVIEW

**Comment by r30shah** [comment]
@fjeremic Can you please review this? I have kept changes in four separate logical commits so that it will be easier to review.
These are the commits you will be interested in, 
1. https://github.com/eclipse/openj9/pull/466/commits/deeb0ea9770b5111db16c0aa921aa4ef64eeaa6b
2. https://github.com/eclipse/openj9/pull/466/commits/3d872200f7cf997e62962140eb919b2c209eb612
3. https://github.com/eclipse/openj9/pull/466/commits/3c3aebf8f87b39ba6d769d8ed876d881be497e68

Please note, For `VMnewEvaluator` and `VMArrayStoreCHK` only we are creating call node as in these case we need to change structure of original node to make it compatible for helper call on IBM Z. For rest of cases, original node is sufficient enough to generate JIT helper call.

**Comment by r30shah** [comment]
@0dvictor @gacholio  I have made changes which Victor has done in https://github.com/eclipse/openj9/pull/141. This is the commit, https://github.com/eclipse/openj9/pull/466/commits/54a2c59bec34e5e801b0c3b0b8b769fad4a39a38
Can I get a review on these changes?

File: runtime/codert_vm/cnathelp.cpp [hunk 1]
```diff
@@ -2491,7 +2491,7 @@ fast_jitCollapseJNIReferenceFrame(J9VMThread *currentThread)
 }
 
 UDATA J9FASTCALL
-#if defined(J9VM_ARCH_X86)
+#if defined(J9VM_ARCH_X86) || defined(J9VM_ARCH_S390)
 /* TODO Will be cleaned once all platforms adopt the correct parameter order */
 fast_jitInstanceOf(J9VMThread *currentThread, j9object_t object, J9Class *castClass)
 #else /* J9VM_ARCH_X86 */
```
Comment (0dvictor) [inline runtime/codert_vm/cnathelp.cpp]: This comment may need be updated as well.

**Comment by 0dvictor** [comment]
The change looks good to me once the comments with `#if defined` are updated.

File: runtime/codert_vm/cnathelp.cpp [hunk 2]
```diff
@@ -2491,12 +2491,12 @@ fast_jitCollapseJNIReferenceFrame(J9VMThread *currentThread)
 }
 
 UDATA J9FASTCALL
-#if defined(J9VM_ARCH_X86)
+#if defined(J9VM_ARCH_X86) || defined(J9VM_ARCH_S390)
 /* TODO Will be cleaned once all platforms adopt the correct parameter order */
 fast_jitInstanceOf(J9VMThread *currentThread, j9object_t object, J9Class *castClass)
-#else /* J9VM_ARCH_X86 */
+#else /* J9VM_ARCH_X86 || J9VM_ARCH_S390*/
```
Comment (r30shah) [inline runtime/codert_vm/cnathelp.cpp]: Added the condition for #if in comment as well.

File: runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp [hunk 3]
```diff
@@ -122,50 +122,6 @@ class S390CHelperLinkage : public TR::Linkage
...
-    *  \return
-    *      Returns TR::Register* which contains return value from helper call
-    */  
-   TR::Register *buildDirectDispatch(TR::Node *callNode, TR::RegisterDependencyConditions** deps, TR::Register *returnReg=NULL)
-   	{
-   	TR_Stack<TR::Register*> paramInRegisters(cg()->trMemory());
-   	return buildDirectDispatch(callNode, deps, paramInRegisters, returnReg);
-   	}
-   TR::Register *buildDirectDispatch(TR::Node *callNode, TR::RegisterDependencyConditions** deps, TR_Stack<TR::Register*>& paramInRegisters, TR::Register *returnReg);
+   TR::Register *buildDirectDispatch(TR::Node *callNode, TR::RegisterDependencyConditions** deps, TR::Register *returnReg=NULL);
```
Comment (fjeremic) [inline runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp]: Do you think reordering `deps` and `returnReg` would be more intuitive given that we have an API above defined as `buildDirectDispatch(TR::Node *callNode, TR::Register *returnReg)`? The order of parameters may then be consistent.

Comment (r30shah) [reply runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp]: Thanks for pointing this out, It does look more intuitive the way you said.  Let me change it. 
Currently we are encountering following uses of the interface. 
```
1. buildDirectDispatch(node) -> Standard API
2. buildDirectDispatch(node, register) -> If result register is already allocated
3. buildDirectDispatch(node, &deps, register=NULL) -> It is called within ICF and can be used same if we have already allocated register)
```

Comment (r30shah) [reply runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp]: Ok, So checked how it looks if we change the order
```
TR::Register* buildDirectDispatch(TR::Node *node, TR::Register *register, TR::RegisterDependencyConditions ** deps = static_cast<TR::RegisterDependencyConditions**>(NULL));

TR::Register * buildDirectDispatch(TR::Node *node)
{
return buildDirectDispatch(node, NULL, static_cast<TR::RegisterDependencyConditions**>(NULL));
}
TR::Register *buildDirectDispatch(TR::Node *node, TR::RegisterDependencyConditions **deps)
{
return buildDirectDispatch(node, NULL, deps);
}
``` 
Even in this way we are ending up with 3 overridden functions. I doubt if this makes that more difference.

Comment (fjeremic) [reply runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp]: @r30shah I understand we cannot get away without three overridden functions. All I was pointing towards was that if we have possible permutation of three parameters `A`, `B`, and `C` defining three APIs as:

```
foo(A);
foo(A,B);
foo(A,B,C);
```

May be more intuitive for the user looking for the appropriate API to use than what we have now, which is:

```
foo(A);
foo(A,B);
foo(A,C,B);
```

i.e. `B` is not consistently declared as the second parameter in two of the APIs. Defining a consistent parameter order may help the user to look for a specific API as the APIs evolve and grow. Take a look at `generateS390MemoryReference` as an example here:

https://github.com/eclipse/omr/blob/master/compiler/z/codegen/OMRMemoryReference.hpp#L461-L474

The parameter ordering is not consistent at all. The `cg` parameter sometimes comes first, sometimes second, sometimes last, etc and there are 13 such overloads. Do we believe a user may have an easier time finding the right API to use if all the parameters were consistently ordered?

Comment (r30shah) [reply runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp]: @fjeremic I got your point. And what we have `generateS390MemoryReference` is really inconsistent and I agree with the point that user would have difficult time to find correct signature to call. 
I would say this case is little different as neither `TR::Register*` nor `TR::RegisterDependencyConditons**` is consistently second parameter. 
I want to be able to call 
`buildDirectDispatch(node)`
`buildDirectDispatch(node, &deps)`
`buildDirectDispatch(node, register)`
`buildDirectDispatch(node, &deps, register)`
And the way I achieve this is by declaring function like,
`TR::Register * buildDirectDispatch(TR::Node *node, TR::RegisterDependencyConditions **deps, TR::Register *reg=NULL`
This was I can call the same function `buildDirectDispatch(node, &deps)` with reg defined `NULL`.
And Declared `TR::Register *buildDirectDispatch(TR::Node *node, TR::Register *reg)` is used to call dispatch function without dependency conditions.

Comment (fjeremic) [reply runtime/tr.source/trj9/z/codegen/J9S390CHelperLinkage.hpp]: I see, so there are four permutations. I have overlooked the fact that in `buildDirectDispatch(node, register)` the `register` parameter does not have a `NULL` default. Thanks for the explanation @r30shah!

**Comment by fjeremic** [comment]
Jenkins test sanity all

**Pull request merged**